ui/admin/app/components/form/policy/index.js

```diff
@@ -0,0 +1,172 @@
+  }
+
+  /**
+   * Returns policy type
+   * @type {string}
+   */
+  get selectRetentionPolicyType() {
+    if (this.retain_days < 0) {
+      return 'forever';
+    } else if (this.retain_days >= 1) {
+      return 'custom';
+    } else {
+      return 'do_not_protect';
+    }
+  }
+
+  /**
+   * Returns policy type
+   * @type {string}
+   */
+  get selectDeletePolicyType() {
+    return this.delete_days > 0 && this.is_custom_deletion_selected
+      ? 'custom'
+      : 'do_not_delete';
+  }
+
+  //actions
+  /**
+   * Show custom text field when custom option is selected
+   */
+
+  @action
+  handlePolicyTypeSelection({ target: { value: selectedVal, name: policy } }) {
+    switch (policy) {
+      case 'retention_policy':
+        this.retain_days = selectedVal;
+        //When `forever` is selected, we should automatically select `do_not_delete`
+        //and hide the custom input
+        if (selectedVal < 0) {
+          this.delete_days = 0;
+          this.is_custom_deletion_selected = false;
+        }
+        if (selectedVal > 0) {
+          this.is_custom_retention_selected = true;
+        } else {
+          this.is_custom_retention_selected = false;
+        }
+        break;
+
+      case 'deletion_policy':
+        this.delete_days = selectedVal;
+        if (selectedVal > 0) {
+          this.is_custom_deletion_selected = true;
+        } else {
+          this.is_custom_deletion_selected = false;
+        }
+        break;
+    }
+  }
+
+  /**
+   * Toggle retain for & delete after overridables
+   */
+  @action
+  handleOverridableToggle(field) {
+    this[field] = !this[field];
+  }
+
+  /**
+   * Set retain for input value to a tracked property
+   */
+  @action
+  handleRetainForInput({ target: { value } }) {
+    this.retain_days = value;
+  }
+
+  /**
+   * Set delete after input value to a tracked property
+   */
+  @action
+  handleDeleteAfterInput({ target: { value } }) {
+    this.delete_days = value;
+  }
+
+  /**
+   * Callback submit and construct retain_days and delete_days model objects
+   */
+  @action
+  async submit() {
+    this.args.model.retain_for = {
+      days: this.retain_days,
+      overridable: this.retain_for_overridable,
+    };
+    this.args.model.delete_after = {
+      days: this.delete_days,
+      overridable: this.delete_after_overridable,
+    };
+    await this.args.submit();
+  }
+
+  /**
+   * Callback cancel and
+   * Resets all the tracked properties after rollback
+   */
+  @action
+  cancel() {
+    this.args.cancel();
+    this.is_custom_retention_selected = this.args.model.retain_for?.days > 0;
+    this.is_custom_deletion_selected = this.args.model.delete_after?.days > 0;
+    this.retain_for_overridable = this.args.model.retain_for?.overridable;
+    this.delete_after_overridable = this.args.model.delete_after?.overridable;
+    this.retain_days = this.args.model.retain_for?.days;
+    this.delete_days = this.args.model.delete_after?.days;
```

Comment:
Maybe I misunderstood this, but since this is an object type in which `ember-data` doesn't roll back, I had to store these in tracked properties and then construct the model object at the end. 

Wouldn't the transform only be reached when we hit submit, not during the editing state?

Reply:
I meant I thought ember should roll back to its previous reference which was your original tracked object, not that it would transform again

Reply:
You are right; it does maintain previous references. I believe I did not use the models directly in the template earlier, which made me think the objects weren't getting the previous references.